config/initializers/devise.rb

```diff
@@ -13,6 +13,27 @@
   config.attribute_map = { webauth_groups: 'WEBAUTH_LDAPPRIVGROUP' }
 end
 
+if Rails.env.production?
+  # Use Warden callbacks to store and clear user data from Shibboleth since there
+  # aren't any appropriate callbacks in Devise
+  # https://github.com/heartcombo/devise/wiki/Callbacks
+  Warden::Manager.after_authentication do |user, auth, opts|
+    auth.session(:user)['suAffiliation'] = auth.env['suAffiliation']
+    user.affiliations = auth.env['suAffiliation']
+    # Reset EDS session token so that a new session is established
+    auth.env['rack.session']['eds_session_token'] = nil
+  end
+
+  Warden::Manager.after_fetch do |user, auth, opts|
+    user.affiliations = auth.session(:user)['suAffiliation']
+  end
```

Comment:
why do we set `user.affiliations` in both these callbacks? warden docs don't explain what `after_fetch` does; wondering when & in what context it is called